File: desktop/ArgonMac/Sources/ArgonMac/ArgonActivity.swift
import Foundation

/// When this Mac is allowed to poll Argon.
///
/// Reads and writes the same `~/.config/argon/activity.json` the Übersicht
/// readout honours, so pausing from the menu bar pauses everything rather than
/// one of two things that are both polling.
///
/// The saving is smaller than it looks and worth being honest about: a paused
/// refresh of the Python readout still costs an interpreter start, about 0.05s
/// of the 0.06s an active one costs. Pausing stops *this* app's polling
/// completely, because it is one long-lived process — the readout's cost is the
/// spawn, which is why its refresh interval matters more than its gate.
struct ArgonActivity: Equatable {
  var activeFrom: String
  var activeTo: String
  var pausedUntil: Date?
  var scheduleEnabled: Bool

  static let path = FileManager.default.homeDirectoryForCurrentUser
    .appendingPathComponent(".config/argon/activity.json")

  static let defaultFrom = "16:00"
  static let defaultTo = "07:30"

  static func load() -> ArgonActivity {
    let object =
      (try? Data(contentsOf: path))
      .flatMap { try? JSONSerialization.jsonObject(with: $0) as? [String: Any] } ?? [:]

    var paused: Date?
    if let stamp = object["paused_until"] as? String {
      paused = Self.parseStamp(stamp)
    }
    return ArgonActivity(
      activeFrom: (object["active_from"] as? String) ?? defaultFrom,
      activeTo: (object["active_to"] as? String) ?? defaultTo,
      pausedUntil: paused,
      scheduleEnabled: (object["schedule_enabled"] as? Bool) ?? true
    )
  }

  func save() {
    var object: [String: Any] = [
      "active_from": activeFrom,
      "active_to": activeTo,
      "schedule_enabled": scheduleEnabled,
    ]
    object["paused_until"] =
      pausedUntil.map { ISO8601DateFormatter.argonLocal.string(from: $0) } ?? NSNull()

    guard let data = try? JSONSerialization.data(withJSONObject: object, options: [.prettyPrinted])
    else { return }
    try? FileManager.default.createDirectory(
      at: Self.path.deletingLastPathComponent(), withIntermediateDirectories: true)
    try? data.write(to: Self.path, options: .atomic)
  }

  /// Whether to poll, and what to say when not.
  func status(at now: Date = Date()) -> (active: Bool, reason: String) {
    if let pausedUntil, pausedUntil > now {
      return (false, "Paused until \(Self.shortTime.string(from: pausedUntil))")
    }
    guard scheduleEnabled else { return (true, "") }

    let minutes = Self.minutes(of: now)
    guard let start = Self.minutes(hhmm: activeFrom), let end = Self.minutes(hhmm: activeTo)
    else {
      return (true, "")
    }
    // 16:00–07:30 wraps past midnight, which is the normal case here rather
    // than the edge case — a plain `start <= now < end` is false all evening.
    let inside = start == end || (start < end
      ? (minutes >= start && minutes < end)
      : (minutes >= start || minutes < end))
    return inside ? (true, "") : (false, "Asleep until \(activeFrom)")
  }

  /// Parse a stamp written by either side.
  ///
  /// Python now writes an offset so `ISO8601DateFormatter` can read it, but a
  /// file written before that is naive local time, which it refuses outright.
  /// Failing to parse would read as "not paused" and quietly resume polling.
  static func parseStamp(_ value: String) -> Date? {
    if let date = ISO8601DateFormatter.argonLocal.date(from: value) { return date }
    let naive = DateFormatter()
    naive.calendar = Calendar(identifier: .gregorian)
    naive.locale = Locale(identifier: "en_US_POSIX")
    naive.timeZone = .autoupdatingCurrent
    naive.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
    return naive.date(from: value)
  }

  private static func minutes(of date: Date) -> Int {
    let parts = Calendar.current.dateComponents([.hour, .minute], from: date)
    return (parts.hour ?? 0) * 60 + (parts.minute ?? 0)
  }

  private static func minutes(hhmm: String) -> Int? {
    let parts = hhmm.split(separator: ":")
    guard parts.count == 2, let h = Int(parts[0]), let m = Int(parts[1]) else { return nil }
    return h * 60 + m
  }

  private static let shortTime: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "h:mm a"
    return formatter
  }()
}

extension ISO8601DateFormatter {
  /// Matches what Python's `datetime.isoformat()` writes: local time, offset,
  /// no fractional seconds.
  static let argonLocal: ISO8601DateFormatter = {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime]
    return formatter
  }()
}

File: desktop/argon_activity.py
#!/usr/bin/env python3
"""When the desktop readouts are allowed to touch the network.

One file, honoured by every surface on this Mac — the Übersicht widgets, the
SwiftBar plugin and the native menu bar app — so pausing means pausing, not
pausing one of three things that are all polling.

Why this exists: the readout costs about 0.06s of CPU per refresh, and at two
widgets on a five second timer plus a menu bar plugin that is 43,200 refreshes
a day, roughly 43 minutes of one core. The raw number understates it, because
waking the CPU every five seconds is what keeps a laptop out of the deep idle
states where battery is actually saved.

Stdlib only and Python 3.9-compatible, for the same reason as the readout:
SwiftBar runs plugins from launchd, whose PATH finds /usr/bin/python3.
"""

from __future__ import annotations

import json
from datetime import datetime, time as dtime, timedelta
from pathlib import Path

ACTIVITY_PATH = Path.home() / ".config" / "argon" / "activity.json"

#: Argon's day runs from the end of school to the start of the next one. He is
#: not using it in class, and a laptop polling through the school day is paying
#: for a readout nobody reads.
DEFAULT_ACTIVE_FROM = "16:00"
DEFAULT_ACTIVE_TO = "07:30"


def _parse_hhmm(value, fallback):
    try:
        hour, minute = str(value).split(":")
        return dtime(int(hour), int(minute))
    except (ValueError, AttributeError):
        return fallback


def load(path=None):
    """The current settings, with defaults filled in. Never raises."""
    path = path or ACTIVITY_PATH
    try:
        data = json.loads(Path(path).read_text())
        if not isinstance(data, dict):
            data = {}
    except (OSError, ValueError):
        data = {}
    return {
        "active_from": data.get("active_from") or DEFAULT_ACTIVE_FROM,
        "active_to": data.get("active_to") or DEFAULT_ACTIVE_TO,
        "paused_until": data.get("paused_until"),
        "schedule_enabled": data.get("schedule_enabled", True),
    }


def save(settings, path=None):
    path = Path(path or ACTIVITY_PATH)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(settings, indent=2, sort_keys=True) + "\n")


def in_window(now, start, end):
    """Is *now* inside a daily window that may wrap past midnight?

    16:00–07:30 is the normal case here, so the wrap is the rule rather than the
    edge case: a naive ``start <= now < end`` is false all evening, which would
    have paused Argon during precisely the hours it exists for.
    """
    if start == end:
        return True
    if start < end:
        return start <= now < end
    return now >= start or now < end


def status(now=None, path=None):
    """``(active, reason)`` — whether to fetch, and what to say if not."""
    now = now or datetime.now()
    settings = load(path)

    paused_until = settings.get("paused_until")
    if paused_until:
        try:
            until = datetime.fromisoformat(paused_until)
            # Stamps are written with an offset so Swift can read them, but a
            # file from before that change is naive, and comparing the two
            # raises rather than returning False. Normalise to naive local.
            if until.tzinfo is not None:
                until = until.astimezone().replace(tzinfo=None)
            if until > now:
                return False, "Paused until " + until.strftime("%-I:%M %p")
        except ValueError:
            pass  # Unreadable stamp: treat as not paused rather than stuck off.

    if not settings.get("schedule_enabled", True):
        return True, ""

    start = _parse_hhmm(settings["active_from"], dtime(16, 0))
    end = _parse_hhmm(settings["active_to"], dtime(7, 30))
    if in_window(now.time(), start, end):
        return True, ""
    return False, "Asleep until " + start.strftime("%-I:%M %p")


def pause(minutes=None, path=None):
    """Stop fetching. ``minutes=None`` means until explicitly resumed."""
    settings = load(path)
    # A century is "indefinite" without needing a second representation for it,
    # and it still shows a real date if anything ever prints the raw file.
    delta = timedelta(minutes=minutes) if minutes else timedelta(days=36500)
    # astimezone() rather than a bare now(): a naive stamp has no offset, and
    # Swift's ISO8601DateFormatter refuses to parse one — so a pause set from
    # the desktop widget would be silently invisible to the menu bar app, which
    # would carry on polling while claiming to be paused.
    settings["paused_until"] = (
        (datetime.now() + delta).astimezone().replace(microsecond=0).isoformat()
    )
    save(settings, path)
    return settings


def resume(path=None):
    settings = load(path)
    settings["paused_until"] = None
    save(settings, path)
    return settings


def selftest():
    import tempfile

    with tempfile.TemporaryDirectory() as tmp:
        path = Path(tmp) / "activity.json"

        # The overnight window is the whole point; a naive comparison fails here.
        assert in_window(dtime(18, 0), dtime(16, 0), dtime(7, 30)) is True
        assert in_window(dtime(2, 0), dtime(16, 0), dtime(7, 30)) is True
        assert in_window(dtime(7, 29), dtime(16, 0), dtime(7, 30)) is True
        assert in_window(dtime(7, 30), dtime(16, 0), dtime(7, 30)) is False
        assert in_window(dtime(11, 0), dtime(16, 0), dtime(7, 30)) is False
        assert in_window(dtime(15, 59), dtime(16, 0), dtime(7, 30)) is False

        # Default schedule: quiet during school, awake after it.
        assert status(datetime(2026, 8, 17, 10, 0), path)[0] is False
        assert status(datetime(2026, 8, 17, 16, 30), path)[0] is True
        assert status(datetime(2026, 8, 17, 23, 0), path)[0] is True

        # Pausing beats the schedule, even inside the active window. Checked
        # against the real clock, because pause() stamps the real one — asking
        # about a fake date two days out just reads as an expired pause.
        inside = datetime.now().replace(hour=17, minute=0, second=0, microsecond=0)
        pause(30, path)
        assert status(inside, path)[0] is False
        resume(path)
        assert status(inside, path)[0] is True

        # Indefinite pause: no argument, still off tomorrow.
        pause(None, path)
        assert status(inside + timedelta(days=1), path)[0] is False
        resume(path)

        # An expired pause is simply over. Both stamp and clock come from the
        # same instant here — pinning "now" to a fixed hour while stamping the
        # real one is how this assertion accidentally tested the opposite.
        moment = datetime.now().replace(hour=17, minute=0, second=0, microsecond=0)
        save({"paused_until": (moment - timedelta(minutes=1)).isoformat()}, path)
        assert status(moment, path)[0] is True

        # A naive stamp still parses: files written before offsets were added.
        save({"paused_until": (moment + timedelta(minutes=5)).isoformat()}, path)
        assert status(moment, path)[0] is False

        # And an offset-aware one, which is what pause() writes now.
        aware = (moment + timedelta(minutes=5)).astimezone().isoformat()
        save({"paused_until": aware}, path)
        assert status(moment, path)[0] is False

        # Schedule off means always awake.
        save({"schedule_enabled": False}, path)
        assert status(datetime(2026, 8, 17, 10, 0), path)[0] is True
    print("ok")


if __name__ == "__main__":
    selftest()
